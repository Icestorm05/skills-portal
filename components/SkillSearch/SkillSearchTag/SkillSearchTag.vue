<template lang="pug">
    v-chip(:style="{ backgroundColor }"
           :text-color="color"
           :close="close"
           class="chip--select-multi"
           @input="$emit('close-tag')").SkillSearchTag
        strong.SkillSearchTagText {{ text }}
          slot
</template>

<script lang="ts">
import Color from "color";
import Component from "nuxt-class-component";
import Vue from "vue";
import { Prop } from "vue-property-decorator";

import CalcTextColor from "~/mixins/CalcTextColor";

import { ISkillCategoryLegend } from "types/legend";

import VChip from "vuetify/es5/components/VChip";

interface IAny {
  [key: string]: string | string[];
}

@Component({
  components: {
    VChip
  },
  mixins: [CalcTextColor]
})
export default class SkillSearchTag extends CalcTextColor {
  @Prop({ type: Array })
  private autocompleteItems: Array<IAny | string> | undefined;
  @Prop({ type: String })
  private autocompleteItemTagsKey: string | undefined;
  @Prop({ type: Boolean, default: false })
  private close: boolean;
  @Prop({ type: Object })
  private legend: ISkillCategoryLegend | undefined;
  @Prop({ type: String })
  private text: string | undefined;
  @Prop({ type: Array })
  private tags: string[] | string | undefined;
  @Prop({ type: Boolean, default: false })
  private prioritizeParentColor: boolean;

  private get backgroundColor() {
    const tags = !this.tags && this.text ? this.computedTags : this.tags;
    const key = this.text ? this.getKey(this.text) : undefined;
    const parentKey = Array.isArray(tags)
      ? this.getKey(tags[tags.length - 1])
      : undefined;

    if (this.prioritizeParentColor) {
      if (!key && parentKey && !this.tags && this.legend) {
        return Color(this.legend[parentKey])
          .lighten(0.2)
          .hex();
      } else if ((key || parentKey) && this.legend) {
        return this.legend[parentKey ? parentKey : (key as string)];
      } else {
        return "lightgrey";
      }
    } else {
      if (!key && parentKey && this.legend) {
        return Color(this.legend[parentKey])
          .lighten(0.2)
          .hex();
      } else if ((key || parentKey) && this.legend) {
        return this.legend[parentKey ? parentKey : (key as string)];
      } else {
        return "lightgrey";
      }
    }
  }

  private get color() {
    const text = this.textColor(this.backgroundColor);
    return text === "#000000" ? "black" : "white";
  }

  private get computedTags() {
    if (this.autocompleteItems && this.autocompleteItemTagsKey) {
      const autocompleteItem = this.autocompleteItems.find(item => {
        const tags =
          typeof item !== "string"
            ? item[this.autocompleteItemTagsKey as string]
            : undefined;
        return Array.isArray(tags) && this.text
          ? tags.includes(this.text)
          : false;
      });
      return typeof autocompleteItem === "object"
        ? autocompleteItem[this.autocompleteItemTagsKey]
        : undefined;
    }
  }

  private getKey(category: string) {
    if (this.legend) {
      return Object.keys(this.legend).find(
        colorKey =>
          this.replaceAnds(colorKey).toLowerCase() ===
          this.replaceAnds(category).toLowerCase()
      );
    } else {
      return "lightgrey";
    }
  }

  private replaceAnds(text: string) {
    return text.replace(/&|and/gi, "&");
  }
}
</script>

<style lang="stylus" scoped>
.SkillSearchTag {
    max-width: 100%;
    /deep/ .chip__content {
      cursor: pointer;
      width: 100%;
    }
    .SkillSearchTagText {
      overflow: hidden;
      text-overflow: ellipsis;
    }
}
</style>
