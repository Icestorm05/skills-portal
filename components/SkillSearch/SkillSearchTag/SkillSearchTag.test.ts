import SkillSearchTag from "./SkillSearchTag.vue";
import { mount } from "@vue/test-utils";

declare const expect: jest.Expect;

describe("Skill Search", () => {
  test("is a Vue instance", () => {
    const wrapper = mount(SkillSearchTag);
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test("snapshot has not changed", async () => {
    const wrapper = mount(SkillSearchTag);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
