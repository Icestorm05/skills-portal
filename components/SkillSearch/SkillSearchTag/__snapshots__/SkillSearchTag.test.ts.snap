// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skill Search snapshot has not changed 1`] = `<span tabindex="0" class="chip chip--select-multi SkillSearchTag black--text"><span class="chip__content"><strong class="SkillSearchTagText"></strong></span></span>`;
